codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,348 @@
+                        public void run() {
+                            streamInfo.childChannel.streamTerminated = true;
+                            streamInfo.childChannel.unsafe().disconnect(
+                                streamInfo.childChannel.newPromise());
+                        }
+                    });
+                }
+            }
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayMessage goAway = new Http2GoAwayMessage(errorCode, debugData);
+            try {
+                http2Handler.connection().forEachActiveStream(new Http2StreamVisitor() {
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                && http2Handler.connection().local().isValidStreamId(stream.id())) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                            if (streamInfo != null) {
+                                streamInfo.childChannel.pipeline()
+                                    .fireUserEventTriggered(goAway.retain());
+                            }
+                        }
+                        return true;
+                    }
+                });
+            } catch (Throwable t) {
+                ctx.invoker().invokeExceptionCaught(ctx, t);
+            }
+            ctx.fireUserEventTriggered(goAway);
+        }
+    }
+
+    class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
+        public InternalHttp2ConnectionHandler(Http2Connection connection,
+                Http2FrameReader frameReader, Http2FrameWriter frameWriter,
+                Http2FrameListener listener) {
+            super(connection, frameReader, frameWriter, listener);
+        }
+
+        @Override
+        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                Http2Exception.StreamException http2Ex) {
+            super.onStreamError(ctx, cause, http2Ex);
+            Http2Stream stream = http2Handler.connection().stream(http2Ex.streamId());
+            if (stream == null) {
+                return;
+            }
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            if (streamInfo == null) {
+                return;
+            }
+            streamInfo.childChannel.pipeline().fireExceptionCaught(http2Ex);
+        }
+    }
+
+    class FrameListener extends Http2FrameAdapter {
+        @Override
+        public void onRstStreamRead(ChannelHandlerContext ctx, int streamId, long errorCode)
+                throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().stream(streamId);
+            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+            // Use a user event in order to circumvent read queue.
```

Comment:
I guess all of this is fine....my mindset has been these are frame events that are read from the socket.  It would be helpful to list out what the expected method calls are w.r.t these events.  For example when a RST_STREAM is read for a given stream...what is the expected sequence of events?  Is the user expected to call "close"...when will the user event be dispatched relative to channelInactive notificaitons vs close listener notifications etc...